Waiting for healthy container when HEALTCHECK is defined

kfrajtak opened the pull request:
## What does this PR do?

Adds the ability to wait till the container is in healthy state reported through health status.

## Why is it important?

In some situations, the container is in a running state, but the containerized application is not ready. HEALTHCHECK probe can be used to detect the state when application is ready - the healthy state.

## How to test this PR

Tests have been added to verify the behaviour.

kfrajtak committed:
- added Health property to expose health status from container
- added new wait (UntilContainerIsHealthy) to wait for healthy status
- added tests with apropriate files to build containers

kfrajtak committed:
Merge branch 'develop' of github.com:kfrajtak/testcontainers-dotnet into develop

kfrajtak commented:
The build is failing because of 
```
Expected: SortedSet<String> ["Dockerfile", "setup/setup.sh"]
Actual:   SortedSet<String> ["Dockerfile", "healthWaitStrategy/fail/.dockerignore", "healthWaitStrategy/fail/Dockerfile", "healthWaitStrategy/fail/empty_loop.sh", "healthWaitStrategy/ok/.dockerignore", ...]
  Stack Trace:
     at DotNet.Testcontainers.Tests.Unit.ImageFromDockerfileTest.DockerfileArchiveTar() in /home/runner/work/testcontainers-dotnet/testcontainers-dotnet/tests/Testcontainers.Tests/Unit/Images/ImageFromDockerfileTest.cs:line 39
```

How can I fix that? The additional files are part of tests involving health check.
Thanks

HofmeisterAn commented on src/Testcontainers/Containers/IDockerContainer.cs:
```diff
@@ -136,6 +137,14 @@ public interface IRunningDockerContainer
+    /// <value>
+    /// Returns the Docker container health.
+    /// </value>
+    Health Health { get; }
```
Can we wrap it in our own type? We do not expose any `Docker.DotNet.Models` types yet. I would like to keep it consistent (similar to `TestcontainersStates`) for now.

https://github.com/testcontainers/testcontainers-dotnet/blob/2d04584618b544edea0c38d3c2ecfaa03210466c/src/Testcontainers/Containers/TestcontainersContainer.cs#L126-L133

kfrajtak replied:
sure

HofmeisterAn commented on src/Testcontainers/Configurations/WaitStrategies/UntilContainerIsHealthy.cs:
```diff
@@ -0,0 +1,36 @@
+    public Task<bool> Until(ITestcontainersContainer testcontainers, ILogger logger)
+    {
+      var health = testcontainers.Health;
+      if (testcontainers.State != TestcontainersStates.Running || health == null)
```
The collection of wait strategies contains always a strategy (the first one) that checks if the container is running. I don't think that part is necessary.

kfrajtak replied:
fixed

HofmeisterAn commented on src/Testcontainers/Configurations/WaitStrategies/UntilContainerIsHealthy.cs:
```diff
@@ -0,0 +1,36 @@
+        return Task.FromResult(false);
+      }
+
+      if (health.FailingStreak > this.failingStreak)
```
I think we can combine this part with `UntilOperationIsSucceeded`. What about:

```csharp
internal class UntilContainerIsHealthy : UntilOperationIsSucceeded
{
  public UntilContainerIsHealthy(int maxCallCount)
    : base(() => false, maxCallCount)
  {
  }

  public async Task<bool> Until(ITestcontainersContainer testcontainers, ILogger logger)
  {
    _ = await base.Until(testcontainers, logger)
      .ConfigureAwait(false);

    return TestcontainersHealths.Healthy.Equals(testcontainers.Health);
  }
}
```

kfrajtak replied:
I am afraid this won't work without further refactoring - the execution will fail at the first `Until` and won't reach the health check. Your idea was to perform the check the same way as in `UntilOperationIsSucceeded` stragegy, right? That would work if the `operation` was not `Func<bool>` but `Func<ITestcontainersContainer, bool>` instead since I need to pass context to the check function.

HofmeisterAn replied:
> Your idea was to perform the check the same way as in UntilOperationIsSucceeded stragegy, right?

Yes. I do not remember anymore. I think I tested it.

HofmeisterAn replied:
Got it now. Thanks!

HofmeisterAn commented on tests/Testcontainers.Tests/Assets/healthWaitStrategy/fail/.dockerignore:
```diff
@@ -0,0 +1,3 @@
+credHelpers
```
Is this `.dockerignore` necessary?

kfrajtak replied:
file was removed

HofmeisterAn commented on tests/Testcontainers.Tests/Assets/healthWaitStrategy/ok/.dockerignore:
```diff
@@ -0,0 +1,3 @@
+credHelpers
```
Is this `.dockerignore` necessary?

kfrajtak replied:
file was removed

HofmeisterAn commented on tests/Testcontainers.Tests/Unit/Configurations/WaitUntilContainerIsHealthyTest.cs:
```diff
@@ -0,0 +1,45 @@
+    [Fact]
+    public async Task StartsOnceHealthy()
+    {
+      var imageName = await new ImageFromDockerfileBuilder()
```
Can we pass the image as class fixture? We can use the same image to test *fail* and *ok*, by overriding the entrypoint or set an env variable to wait until / write the file?

HofmeisterAn reviewed (changes requested):
Thanks for the PR 🙏 I just have some small suggestions to keep the implementation more consistent. Great feature, looking forward to merge it.

HofmeisterAn commented:
> The build is failing because of
> 
> ```
> Expected: SortedSet<String> ["Dockerfile", "setup/setup.sh"]
> Actual:   SortedSet<String> ["Dockerfile", "healthWaitStrategy/fail/.dockerignore", "healthWaitStrategy/fail/Dockerfile", "healthWaitStrategy/fail/empty_loop.sh", "healthWaitStrategy/ok/.dockerignore", ...]
>   Stack Trace:
>      at DotNet.Testcontainers.Tests.Unit.ImageFromDockerfileTest.DockerfileArchiveTar() in /home/runner/work/testcontainers-dotnet/testcontainers-dotnet/tests/Testcontainers.Tests/Unit/Images/ImageFromDockerfileTest.cs:line 39
> ```
> 
> How can I fix that? The additional files are part of tests involving health check. Thanks

Either we add `healthWaitStrategy` to [.dockerignore](https://github.com/testcontainers/testcontainers-dotnet/blob/develop/tests/Testcontainers.Tests/Assets/.dockerignore) or move it into a subdirectory as you did with `healthWaitStrategy`.

kfrajtak committed:
- failingStreak type changed from int to long
- introduced TestcontainersHealthStates enum as a wrapper to hide Docker.DotNet.Models type
- added two new properties to expose health state and health check failing streak

kfrajtak committed:
unnecessary files removed

kfrajtak committed:
using one Dockerfile with ARG

kfrajtak committed:
early exit when container is not running

kfrajtak committed:
ignore healthWaitStrategy directory

kfrajtak commented:
The pipeline is not running the test report publishing step when previous step fails.
